### ISSUE TITLE
Timing hour problem

### ISSUE THREAD

**Issue by Ciloe** [kelektiv/node-cron#163]
Hi, 

I have a problem with a sequence. I use this : 

``` javascript
var toDo = function () {
        "use strict";
        console.log('Start cron Test');
        console.log('Check hour ', new Date().toISOString());
    },
    CronJob = require('cron').CronJob,
    cronJobTest = new CronJob({
        cronTime: '0 5,10,15,25,59 9,10,21,22 * * *', // At H+5, H+10, H+15, H+25 et H+59 minutes for 9h, 10h, 21h et 22h UTC.
        onTick: toDo,
        start: false,
        timeZone: "UTC"
    });

cronJobTest.start();
```

But, when I see the console log, I have this : 

``` bash
Start cron Test
Check hour 2015-04-16T09:25:00.021Z
Start cron Test
Check hour 2015-04-16T09:59:00.029Z
Start cron Test
Check hour 2015-04-16T20:05:03.337Z
Start cron Test
Check hour 2015-04-16T20:10:00.414Z
Start cron Test
Check hour 2015-04-16T20:15:00.448Z
Start cron Test
Check hour 2015-04-16T20:25:00.509Z
Start cron Test
Check hour 2015-04-16T20:58:59.926Z
Start cron Test
Check hour 2015-04-16T20:59:01.011Z
Start cron Test
Check hour 2015-04-16T21:05:00.035Z
Start cron Test
Check hour 2015-04-16T21:10:00.065Z
Start cron Test
Check hour 2015-04-16T21:15:00.095Z
Start cron Test
Check hour 2015-04-16T21:25:00.163Z
Start cron Test
Check hour 2015-04-16T21:59:00.401Z
```

You can see that the UTC cron job launch one hour before the hour I want. After, you can already see a probleme at the time 20:58:59.926Z and 20:59:01.011Z. It's launch two times.....

What is the problem ?

**Comment by Ciloe** [comment]
No ideas about this problem ?

**Comment by cvillemure** [comment]
We also have the problem on our side where the cron run two times in a row, not respecting the cron tab

**Comment by ncb000gt** [comment]
Sorry. I must have missed this in all the email I get. I'll try to take a look tonight.

**Comment by cvillemure** [comment]
I was able to reproduce the bug on Windows with :
* **cron** version 1.0.5
* **time** version 0.11.0 (Modified to use moment-timezone)
* **node** version 0.10.38 (64 bits)

```
var log4js = require('log4js');
var moment = require('moment');
var CronJob = require('cron').CronJob;
var logger = log4js.getLogger("cronBug");

var joblog = [];
var NB_OF_SECONDS_BETWEEN_JOBS = 1;
var NB_OF_MS_BETWEEN_JOBS = NB_OF_SECONDS_BETWEEN_JOBS * 1000;
var THRESHOLD_IN_MS = 100;
var MIN_NB_OF_MS_BETWEEN_JOBS = NB_OF_MS_BETWEEN_JOBS - THRESHOLD_IN_MS;
var MAX_NB_OF_MS_BETWEEN_JOBS = NB_OF_MS_BETWEEN_JOBS + THRESHOLD_IN_MS;
var TIME_FORMAT = "YYYY-MM-DD HH:mm:ss.SSS";

var job = new CronJob({
	cronTime: '*/' + NB_OF_SECONDS_BETWEEN_JOBS + ' * * * * *',
	onTick: function() {
		var currentJobTime = new Date().getTime();
		var lastJobTime = (joblog.length ? joblog[joblog.length-1] : null);
		var timeDifference = (lastJobTime ? currentJobTime - lastJobTime : null);
		  
		joblog.push(currentJobTime);
		//logger.info("cron start");
		
		if(timeDifference && (timeDifference < MIN_NB_OF_MS_BETWEEN_JOBS)){
			logger.error("The cron time was not respected ...");
			logger.error("    Last job    : " + moment(lastJobTime).format(TIME_FORMAT));
			logger.error("    Current job : " + moment(currentJobTime).format(TIME_FORMAT));
			logger.error("    " + timeDifference + "ms is not near the configured " + NB_OF_MS_BETWEEN_JOBS + "ms between jobs");
			exit(1);
		}
		
		var a = 0;
		
		while(a < 1000000000) {
			a++; //We must have some workload for the bug to occur
		}
	},
	start: false,
	timeZone: 'America/Montreal'
});
job.start();
```

The culprit seems to be **new CronDate()** in the method **sendAt** since when the duplicate execution occurs, the method **_getNextDateFrom** is called twice in a row with the same timestamp!

Base on the previous discoveries, i tried the following :
* Run the same test without timezone : *fail again*
* Run the same test without timezone and without the time module : *no crash*

So I assume we made an error in our implementation of **time**.

Upgrading to the latest version (1.0.9 using moment-timezone by default) I was not able to reproduce the bug.  So I will upgrade to the latest version, drop the **time** module and this should solve my issue.

**Comment by ncb000gt** [comment]
Great. Thanks for the update @cvillemure!

**Comment by ncb000gt** [comment]
@Ciloe what version of node and the cron module are you on?

**Comment by Ciloe** [comment]
Hi, thank you to comment this issue. I'm using cron 1.0.0, npm 2.5.1 and node 0.12.1

**Comment by ncb000gt** [comment]
@Ciloe  try using the latest version of the module. See if that helps.

**Comment by jperkelens** [comment]
I'm using 1.0.9 and getting the same issue. My jobs fire twice for every tick.

**Comment by moroccan-dude** [comment]
Same problem here, job triggers twice. It seems as if the problem only occurs when the time gap between the cronjob start and the time the job actually triggers is big. I.e: I tried setting 
case 1:
trigger job 5 PM
I start my service (or run cronjob.start()) at 1PM
--> Job is triggered only once.
case 2:
trigger job 4 AM
I start my service (or run cronjob.start()) at 1PM
--> Job is triggered twice. once at 3:52 AM and a second time at 4:00 AM.

I am on 1.0.9 and nodejs version 0.10.36

**Comment by ncb000gt** [comment]
That's really strange. We do play around with the timeouts for _really long_ timeouts. However, that breaks down to about 2.5 days and case 2 doesn't sound like that's the case. Since we run on a timeout the job times shouldn't change. It seems like there is another job being created/triggered.

I'm assuming you are both still having this issue? @jperkelens @moroccan-dude

**Comment by jperkelens** [comment]
Couldn't tell you, we ended up adding cron support to our scheduling library.

**Comment by ncb000gt** [comment]
@jperkelens cron as in crontab?

**Comment by jperkelens** [comment]
No, we started depending on the [cron-parser](https://github.com/harrisiirak/cron-parser) library and used it to support crontab like scheduling in a library we were using for one-time event scheduling. The cron-parser library is very simple, you might want to look into it for this project, if you're having trouble nailing down this problem.

**Comment by ncb000gt** [comment]
Gotcha. I had looked at it for this and there was a bug in it. I may look
again.

Thanks for the info!

**Comment by jasson15** [comment]
+1
node v4.3.0
npm v3.8.9
cron@1.0.0
```javascript
    new CronJob('0 10 * * * *', function(){
        dosomething(); // this line is usually launched twice at 00:10:00 AM UTC every day
    }, null, true);
```

**Comment by ncb000gt** [comment]
Closing due to age. Can create a new ticket if it comes up again.
